File: ui/src/ipc/events.ts
/**
 * The `player://…` event channel. The UI mirrors playback state from these snapshots rather
 * than polling, so a change made anywhere in the core reaches the UI by one path.
 */
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { MediaInfo, PlaybackState } from "./types";

export const PLAYER_STATE_EVENT = "player://state";
export const MEDIA_OPENED_EVENT = "player://media-opened";

export const onPlayerState = (handler: (state: PlaybackState) => void): Promise<UnlistenFn> =>
  listen<PlaybackState>(PLAYER_STATE_EVENT, (event) => handler(event.payload));

export const onMediaOpened = (handler: (media: MediaInfo) => void): Promise<UnlistenFn> =>
  listen<MediaInfo>(MEDIA_OPENED_EVENT, (event) => handler(event.payload));

File: src-tauri/src/events.rs
//! The `player://…` event channel the UI mirrors playback state from.
//!
//! The UI never polls and never receives pixels — it renders chrome from these snapshots
//! while the native surface underneath draws the video. Emitting is best-effort: a dropped
//! event is logged, never fatal, because the UI can always re-read `get_state`.

use freally_player_core::{MediaInfo, PlaybackState};
use tauri::{AppHandle, Emitter};

/// Emitted after every transport change.
pub const STATE: &str = "player://state";
/// Emitted when a new media item has been opened.
pub const MEDIA_OPENED: &str = "player://media-opened";

pub fn emit_state(app: &AppHandle, state: &PlaybackState) {
    if let Err(e) = app.emit(STATE, state) {
        log::warn!("could not emit {STATE}: {e}");
    }
}

pub fn emit_media_opened(app: &AppHandle, media: &MediaInfo) {
    if let Err(e) = app.emit(MEDIA_OPENED, media) {
        log::warn!("could not emit {MEDIA_OPENED}: {e}");
    }
}
